### PR TITLE
Add doxygen configuration

### PR DESCRIPTION
**Pull request by goto40**
- Doxygen configuration added (see apidoc.md).
- The egenrated API doc contains a search function.
- The idea is to successive add documentation to the code and the underlying architecture:
  - static information can be added via standard doxygen comments (responsibilities of individual classes/functions/modules)
  - dynamic behaviour can be added using the plantuml syntax (e.g. activity diagrams)

### CODE REVIEW

**Comment by eidheim** [comment]
Thank you. I'm busy till Thursday morning, but can look at this then.

File: docs/apidoc.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# juCi++ API doc
+
+## Prerequisites:
+ * doxygen
+ * plantuml (install via apt-get or download from http://de.plantuml.com/; if downloaded either copy the jar file to /usr/bin or set the environment variable PLANTUML_PATH to point to the path containing the jar file)
+
```
Comment (zalox) [inline docs/apidoc.md]: Notes:
* Link to install instructions provided by platuml instead. java also dependency?
* plantuml generates uml, not needed for api-only build.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -141,3 +141,21 @@ target_link_libraries(${project_name} ${global_libraries})
 install(TARGETS ${project_name}
   RUNTIME DESTINATION bin
 )
+
+find_package(Plantuml)
+if(PLANTUML_FOUND)
+#message("plantuml found PLANTUML_JARFILE=${PLANTUML_JARFILE}")
+else(Plantuml_FOUND)
+message("plantuml not found. No UML diagrams are generated. You may define the environment variable PLANTUML_PATH...")
```
Comment (zalox) [inline src/CMakeLists.txt]: What does this do?

File: docs/apidoc.md (hunk 1 above)
Comment (goto40) [inline docs/apidoc.md]: Yes, java is also a dependency to run plantuml. Plantuml allows to specify uml diagrams within doxygen comments.
Without plantuml, the API doc will omit the UML diagrams.

File: src/CMakeLists.txt (hunk 2 above)
Comment (goto40) [inline src/CMakeLists.txt]: with find_package(PlantUml) plantuml is detected and later inserted into the Doxyfile (generated form Doxyfile.in). 
As noted above, if it is not detected UML diagrams will just be omitted.

File: docs/apidoc.md (hunk 1 above)
Comment (goto40) [inline docs/apidoc.md]: I think with ubuntu 16.04 "apt-get install plantuml" just does the thing (I will check this).

File: src/CMakeLists.txt (hunk 2 above)
Comment (zalox) [inline src/CMakeLists.txt]: I see, so doxygen calls plantuml?

Comment (goto40) [inline src/CMakeLists.txt]: Especially behaviour diagrams may help to provide graphical information how different components work together (e.g. using activity diagrams). See: http://goto40.github.io/create_anaglyph/page_architecture.html ("Runtime View")

Comment (goto40) [inline src/CMakeLists.txt]: Yes, doxygen integrates plantuml...

**Comment by eidheim** [comment]
I think this might help new contributors a lot in understanding the source. I'm not a doxygen expert, but can we set HAVE_DOT to YES (https://github.com/goto40/jucipp/blob/add_doxygen_config/src/Doxyfile.in#L2114)? That enabled some other useful diagrams.

**Comment by goto40** [comment]
I added HAVE_DOT=YES and also enabled caller graphs (showing who is calling a particular function).

**Pull request merged**